File: lclang/src/main/java/lclang/statements/expressions/CallExpression.java
package lclang.statements.expressions;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;
import lclang.exceptions.LCLangTypeErrorException;
import lclang.libs.lang.classes.LCClass;
import lclang.methods.Method;
import lclang.types.CallableType;
import lclang.types.Type;
import lclang.utils.TypeUtils;

import java.util.ArrayList;
import java.util.List;

public class CallExpression extends Expression {
    public final Expression expression;
    public final List<Expression> arguments;

    public CallExpression(Expression expression, List<Expression> arguments, int line) {
        super(line);
        this.expression = expression;
        this.arguments = arguments;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        Caller expressionCaller = expression.getCaller(prevCaller);
        LCClass callClazz = expression.visit(prevCaller, visitor).get.invoke(expressionCaller);
        if(!(callClazz.classType instanceof CallableType))
            throw new LCLangTypeErrorException("Value is not callable (it is "+callClazz.classType+")", expressionCaller);

        ArrayList<Type> argsTypes = new ArrayList<>();
        ArrayList<LCClass> args = new ArrayList<>();
        for(Expression argument: arguments) {
            Link argumentValue = argument.visit(prevCaller, visitor);
            LCClass clazz = argumentValue.get.invoke(expression.getCaller(prevCaller));
            argsTypes.add(clazz.classType);
            args.add(clazz);
        }

        Method method = (Method) callClazz;
        if(method.args.size()!=argsTypes.size()){
            throw method.args.size()>argsTypes.size() ?
                    new LCLangTypeErrorException("Invalid arguments: few arguments",
                            getCaller(prevCaller)):
                    new LCLangTypeErrorException("Invalid arguments: too many arguments",
                            arguments.get(method.args.size()).getCaller(prevCaller));
        }

        int notAcceptArg = TypeUtils.isAccept(method.args, argsTypes);
        if(notAcceptArg!=-1)
            throw new LCLangTypeErrorException("Invalid argument "+(notAcceptArg+1)+
                    ": excepted "+method.args.get(notAcceptArg)+", but received "+argsTypes.get(notAcceptArg),
                    arguments.get(notAcceptArg).getCaller(prevCaller));

        return method.call(getCaller(prevCaller), args).createLink();
    }
}

File: lclang/src/main/kotlin/lclang/statements/CallExpression.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.Value
import lclang.exceptions.TypeErrorException
import lclang.isAccept
import lclang.methods.Method
import lclang.types.CallableType
import lclang.types.Type

class CallExpression(
    val expression: Expression,
    val arguments: Array<Expression>,
    line: Int,
    column: Int
): Expression(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        val caller = getCaller(prevCaller)
        val value = expression.visit(caller, visitor)
        if(value.type !is CallableType)
            throw TypeErrorException("Value is not callable (it is ${value.type})", caller)

        val argsTypes = ArrayList<Type>()
        val args = ArrayList<Value>()
        for(argument in arguments) {
            val argumentValue = argument.visit(caller, visitor)
            argsTypes.add(argumentValue.type)

            args.add(argumentValue)
        }

        val method = value.get(caller) as Method
        if(method.args.size!=argsTypes.size){
            throw TypeErrorException(if(method.args.size>argsTypes.size)
                "Invalid arguments: few arguments" else "Invalid arguments: too many arguments", caller)
        }

        val notAcceptArg = method.args.isAccept(argsTypes.toTypedArray())
        if(notAcceptArg!=-1)
            throw TypeErrorException("Invalid argument $notAcceptArg: excepted ${method.args[notAcceptArg]}",
                caller)
//                expressions[notAcceptArg].asCaller(this))

        return Value(method.returnType, method.call(caller, args))
    }
}